internal/app/machined/pkg/runtime/v1alpha1/bootloader/grub/grub.go

```diff
@@ -44,6 +56,79 @@ func NewConfig() *Config {
 	}
 }
 
+// KexecLoad does a kexec using the bootloader.
+func (c *Config) KexecLoad(r runtime.Runtime, disk string) error {
+	return mount.PartitionOp(
+		disk,
+		[]mount.Spec{
+			{
+				PartitionLabel: constants.BootPartitionLabel,
+				FilesystemType: partition.FilesystemTypeXFS,
+				MountTarget:    constants.BootMountPoint,
+			},
+		},
+		func() error {
+			defaultEntry, ok := c.Entries[c.Default]
+			if !ok {
+				return nil
+			}
+
+			kernelPath := filepath.Join(constants.BootMountPoint, defaultEntry.Linux)
+			initrdPath := filepath.Join(constants.BootMountPoint, defaultEntry.Initrd)
+
+			kernel, err := os.Open(kernelPath)
+			if err != nil {
+				return err
+			}
+
+			defer kernel.Close() //nolint:errcheck
+
+			fd := int(kernel.Fd())
+
+			// on arm64 we need to extract the kernel from the zboot image if it's compressed
+			if goruntime.GOARCH == "arm64" {
+				var fileCloser io.Closer
+
+				fd, fileCloser, err = zboot.Extract(kernel)
+				if err != nil {
+					return err
+				}
+
+				defer func() {
+					if fileCloser != nil {
+						fileCloser.Close() //nolint:errcheck
+					}
+				}()
+			}
```

Comment:
This code is shared with sdboot and can be refactored into a new function

Reply:
but it makes more sense to keep it in here only, since it replaces the `fd` and passing it around might be confusing